### PR TITLE
Basic cluster reconciler

### LINKED ISSUE #106: Implement basic cluster actuator
So far, the cluster actutor is not yet implemented, except for basic types.

I would like to see a way to set up Networks, SecrityGroups and Loadbalancers for the cluster. In a first step, a single subnet for all nodes (master and worker) would IMHO be enough. Created Machines should use the created infrastructure.

Todo:
* Create Network with subnet with a configured CIDR
* Create a router connected to a configured external network, as well as the just created subnet
* Create SecurityGroups for master and nodes
* Create a loadbalancer with floating IP

~~LB Floating IP~~ and NetworkID must go to cluster status. SecurityGroups also somehow, but I am currently not sure yet how. Edit: Floating IP not needed because there is an `APIEndpoint` on `ClusterStatus`.

The loadbalancer should be used as the entry point for the apiserver, so we are able to create HA control planes.

If we are done with this, we can update the machine actuator to use the infrastructure.

See also: https://github.com/kubernetes-sigs/cluster-api-provider-aws/blob/master/pkg/cloud/aws/actuators/cluster/actuator.go

### PR DESCRIPTION
**Pull request by chrigl**
**What this PR does / why we need it**:

Implemented basic cluster reconciler. When a `Cluster.NodeCIDR` and a `Cluster.ExternalNetworkID` are configured, the cluster reconciler creates an OpenStack Network with an associated Subnet (this will hold the IP range). Additionally an OpenStack router will be create, and attached to the just created Subnet as well as the `ExternalNetworkID` to get public internet to the VMs. All of the created components are stored in `OpenstackClusterProviderStatus`.

With this, we are able to change the machine actuator to use the Network from `Cluster.Status.ProviderSpec.Network`. It could be still possible to configure networks on Machine bases.

This code does not interfere with current cluster creation. If there is no network configuration on the cluster, the cluster actuator only is a noop as before. 

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #106 

**Special notes for your reviewer**:

In order to test use this image: `chrigl/openstack-cluster-api-controller:latest`

**Release note**:

```release-note

```

### CODE REVIEW

**Comment by chrigl** [comment]
Needs a rebase and minor change when #136 is in.

**Comment by chrigl** [comment]
Huge rebase necessary because of 136. But everything works now.

File: pkg/cloud/openstack/clients/networkservice.go [hunk 1]
```diff
@@ -0,0 +1,278 @@
...
+	"github.com/gophercloud/gophercloud/openstack/networking/v2/extensions/layer3/routers"
+	"github.com/gophercloud/gophercloud/openstack/networking/v2/networks"
+	"github.com/gophercloud/gophercloud/openstack/networking/v2/ports"
+	"github.com/gophercloud/gophercloud/openstack/networking/v2/subnets"
+	"k8s.io/klog"
+	openstackconfigv1 "sigs.k8s.io/cluster-api-provider-openstack/pkg/apis/openstackproviderconfig/v1alpha1"
+)
+
+const (
+	defaultDNS    string = "8.8.8.8"
```
[the remark this answers is not in the document]
Comment (russellb) [reply pkg/cloud/openstack/clients/networkservice.go]: +1, as people may be expecting internal DNS functionality from Neutron

File: pkg/apis/openstackproviderconfig/v1alpha1/types.go [hunk 2]
```diff
@@ -119,6 +127,30 @@ type OpenstackClusterProviderStatus struct {
...
+	// Network contains all information about the created OpenStack Network.
+	// It includes Subnets and Router.
+	Network Network `json:"network,omitempty"`
+}
+
+type Network struct {
+	Name string `json:"name"`
+	ID   string `json:"id"`
+
+	Subnets []Subnet `json:"subnets,omitempty"`
```
[the remark this answers is not in the document]
Comment (chrigl) [reply pkg/apis/openstackproviderconfig/v1alpha1/types.go]: Changed to Subnet

File: pkg/cloud/openstack/clients/networkservice.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (chrigl) [reply pkg/cloud/openstack/clients/networkservice.go]: Done. If we decide to be able to configure DNS, we easily can add a field to clusterspec.

**Pull request merged**